app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/LocalFileTransferActivity.kt

```diff
@@ -0,0 +1,382 @@
+        )
+      }
+      false
+    } else {
+      true
+      // Control reaches here: Either permission granted at install time, or at the time of request
+    }
+  }
+
+  private fun checkExternalStorageWritePermission(): Boolean { // To access and store the zims
+    return if (ContextCompat.checkSelfPermission(
+        this,
+        Manifest.permission.WRITE_EXTERNAL_STORAGE
+      )
+      == PackageManager.PERMISSION_DENIED
+    ) {
+      if (ActivityCompat.shouldShowRequestPermissionRationale(
+          this,
+          Manifest.permission.WRITE_EXTERNAL_STORAGE
+        )
+      ) {
+        alertDialogShower.show(KiwixDialog.StoragePermissionRationale, {
+          ActivityCompat.requestPermissions(
+            this@LocalFileTransferActivity,
+            arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+            PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS
+          )
+        })
+      } else {
+        ActivityCompat.requestPermissions(
+          this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+          PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS
+        )
+      }
+      false
+    } else {
+      true
+      // Control reaches here: Either permission granted at install time, or at the time of request
+    }
+  }
+
+  override fun onRequestPermissionsResult(
+    requestCode: Int,
+    permissions: Array<String>,
+    grantResults: IntArray
+  ) {
+    if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
+      when (requestCode) {
+        PERMISSION_REQUEST_CODE_COARSE_LOCATION -> {
+          Log.e(
+            TAG,
+            "Location permission not granted"
+          )
+          R.string.permission_refused_location.toast(this)
+
+          finish()
+        }
+        PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS -> {
+          Log.e(
+            TAG,
+            "Storage write permission not granted"
+          )
+          R.string.permission_refused_storage.toast(this)
+
+          finish()
+        }
+        else -> {
+          super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        }
+      }
+    }
+  }
+
+  private val isLocationServiceEnabled: Boolean
+    get() = isProviderEnabled(LocationManager.GPS_PROVIDER) ||
+      isProviderEnabled(LocationManager.NETWORK_PROVIDER)
+
+  private fun isProviderEnabled(locationProvider: String): Boolean {
+    return try {
+      locationManager.isProviderEnabled(locationProvider)
+    } catch (ex: SecurityException) {
+      ex.printStackTrace()
+      false
+    } catch (ex: IllegalArgumentException) {
+      ex.printStackTrace()
+      false
+    }
+  }
+
+  private fun requestEnableLocationServices() {
+    alertDialogShower.show(
+      KiwixDialog.EnableLocationServices, {
+        startActivityForResult(
+          Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS),
+          REQUEST_ENABLE_LOCATION_SERVICES
+        )
+      },
+      R.string.discovery_needs_location.toast(this)::show
```

Comment:
Are you sure that the toast is being displayed and no `getString()` is required? As R.string...... usually leads to some int value

Reply:
Oh :facepalm: i didn't notice, sir has already considered that, sorry for pain